### PR TITLE
Add a ConfigParserPlugin to extend config keys

### PR DESCRIPTION
**Pull request by theopolis**
When new components need to store data in the config they should add top-level JSON keys and access the data through a `ConfigDataInstance::data` property tree. If an action should be performed when a configuration is updated or some global state must be tracked then a `ConfigParserPlugin` should be used to "handle" a set of config top-level keys. The parser much implement an update method that receives a potentially-empty property tree for each key requested.

### CODE REVIEW

**Comment by theopolis** [comment]
ok to test

**Comment by theopolis** [comment]
ok to test

**Comment by theopolis** [comment]
@wxsBSD, you may need to resolve conflicts after this. There's some (very tiny) optimizations/changes to YARA code; and the matches utils do not need to touch config.cpp anymore. If you run into any problems merging feel free to create a WIP PR or branch and I can resolve conflicts for you. :)

File: osquery/config/config.cpp [hunk 1]
```diff
@@ -75,6 +75,25 @@ Status Config::update(const std::map<std::string, std::string>& config) {
...
+    auto parser = std::static_pointer_cast<ConfigParserPlugin>(plugin.second);
+    if (parser == nullptr || parser.get() == nullptr) {
+      continue;
+    }
+
+    // For each key requested by the parser, add a property tree reference.
+    std::map<std::string, ConfigTree> parser_config;
+    for (const auto& key : parser->keys()) {
+      if (conf.all_data.count(key) > 0) {
+        parser_config[key] = conf.all_data.get_child(key);
```
Comment (marpaia) [inline osquery/config/config.cpp]: does `get_child` throw if `key` doesn't exist?

Comment (theopolis) [inline osquery/config/config.cpp]: Yes, but the `.count(key)` above makes sure it exists. ;)

Comment (marpaia) [inline osquery/config/config.cpp]: ah, duh :)

**Pull request merged**